cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -763,119 +731,14 @@ private void setServiceInstances(ProgramId programId, int instances)
-
-          // Check if run id is for the right program type
-          if (targetProgramId != null) {
-            runIdToProgramId.put(runId, targetProgramId);
-            return true;
-          } else {
-            return false;
-          }
-        }
-      });
-
-    // don't correct run records for programs running inside a workflow
-    // for instance, a MapReduce running in a Workflow will not be contained in the runtime info in this class
-    invalidRunRecords = Collections2.filter(invalidRunRecords, new com.google.common.base.Predicate<RunRecordMeta>() {
-      @Override
-      public boolean apply(RunRecordMeta invalidRunRecordMeta) {
-        boolean shouldCorrect = shouldCorrectForWorkflowChildren(invalidRunRecordMeta, processedInvalidRunRecordIds);
-        if (!shouldCorrect) {
-          LOG.trace("Will not correct invalid run record {} since it's parent workflow still running.",
-                    invalidRunRecordMeta);
-          return false;
-        }
-        return true;
-      }
-    });
-
-    LOG.trace("End getting invalid run records.");
-
-    if (!invalidRunRecords.isEmpty()) {
-      LOG.warn("Found {} RunRecords with RUNNING status and the program not actually running for program type {}",
-               invalidRunRecords.size(), programType.getPrettyName());
-    } else {
-      LOG.trace("No RunRecords found with RUNNING status and the program not actually running for program type {}",
-                programType.getPrettyName());
-    }
-
-    // Now lets correct the invalid RunRecords
-    for (RunRecordMeta invalidRunRecordMeta : invalidRunRecords) {
-      String runId = invalidRunRecordMeta.getPid();
-      ProgramId targetProgramId = runIdToProgramId.get(runId);
-
-      boolean updated = store.compareAndSetStatus(targetProgramId, runId, ProgramController.State.ALIVE.getRunStatus(),
-                                                  ProgramController.State.ERROR.getRunStatus());
-      if (updated) {
-        LOG.warn("Fixed RunRecord {} for program {} with RUNNING status because the program was not " +
-                   "actually running",
-                 runId, targetProgramId);
-
-        processedInvalidRunRecordIds.add(runId);
-      }
-    }
-  }
-
-  /**
-   * Helper method to check if the run record is a child program of a Workflow
+   * Method to check if the run record is a child program of a Workflow
    *
    * @param runRecordMeta The target {@link RunRecordMeta} to check
    * @param processedInvalidRunRecordIds the {@link Set} of processed invalid run record ids.
    * @return {@code true} of we should check and {@code false} otherwise
    */
-  private boolean shouldCorrectForWorkflowChildren(RunRecordMeta runRecordMeta,
-                                                   Set<String> processedInvalidRunRecordIds) {
+  protected boolean shouldCorrectForWorkflowChildren(RunRecordMeta runRecordMeta,
```

Comment:
Actually, `shouldCorrectForWorkflowChildren` and `retrieveProgramIdForRunRecord` should be moved to `AbstractRunRecordCorrectorService`, since they are only used for correcting run records.